Add e2e tests for attributes count

roykho opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Closes https://github.com/woocommerce/woocommerce-blocks/issues/8873

This does not cover all the tests required from above mentioned issue. It is a starter to gauge if we're going in the right direction as Playwright was just recently added. Open to any suggestions if there are better approaches.

Note that I had to use some hacky solution to get around the generation of the attributes lookup table. 

**Tests covered:**

* [x] Only one attribute filter
* [x] Two different attribute filters
* [x] One attribute filter used alongside the price filter

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. `npm run env:start`
2. `npx playwright install`
3. `npm run test:e2e-pw ./tests/e2e-pw/tests/attributes-filter/filter-products-by-attributes-count.block_theme.spec.ts`
4. Ensure the tests passes.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Add e2e tests for attributes count

roykho committed:
Add e2e tests for attributes count

roykho committed:
Add test with pricing filter and turn on debug to prevent cache

roykho committed:
Merge branch 'trunk' into add/e2e-tests-attributes-count

roykho committed:
Merge branch 'trunk' into add/e2e-tests-attributes-count

nefeline commented on tests/e2e-pw/tests/attributes-filter/filter-products-by-attributes-count.block_theme.spec.ts:
```diff
@@ -0,0 +1,121 @@
+		'.regenerate_product_attributes_lookup_table input'
+	);
+
+	await page.click( '.regenerate_product_attributes_lookup_table input' );
```
Instead of regenerating the product attributes lookup table at this point, how about enabling lookup direct updates? That's what we did in our [pre-existing e2e tests](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/fixtures/fixture-loaders.js#L484-L490) to circumvent the same problem of not having this table populated. 

This setting ensures that the lookup table is immediately populated whenever a new product is created instead of scheduling an action to do so async.

gigitux replied:
If there isn't a way to fix this issue in another way, I would prefer to have this step at the [setup level](https://github.com/woocommerce/woocommerce-blocks/blob/9c33c0819e2de8c1390ee560fdaf35df804d4035/tests/e2e-pw/global-setup.ts).

nefeline reviewed (commented):
Nice work 🎉 !

I ran the tests locally, and they are all passing, but I have a few notes:

- What do you think about updating the name/description for those tests to something shorter so the entire content can be read (note the ellipsis at the beginning of each line)?
- I'm getting a `Failed to load resource: the server responded with a status of 500 (Internal Server Error)` for each one of the tests: are you getting the same?

<img width="1995" alt="Screenshot 2023-05-25 at 23 00 44" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/a4a552e9-d3ad-43a6-a1ce-104773f9fe25">

roykho commented:
@nefeline 

> What do you think about updating the name/description for those tests to something shorter so the entire content can be read (note the ellipsis at the beginning of each line)?

Do you have any suggestions so that it is descriptive enough otherwise if we have a fail somewhere, all the descriptions would be the same and you would have to count to know which failed. I had previously thought about `'Filter by Attributes Block - with All products Block 1...2...3'`. But it decided against it.

> I'm getting a Failed to load resource: the server responded with a status of 500 (Internal Server Error) for each one of the tests: are you getting the same?

No, I am not seeing those 500 errors (sometimes 404s but that happens on other tests too). Are you able to see the PHP logs? Also did you try it with a fresh env? Like `npm run env:restart`? Sometimes it also helps to run `npm run env:stop` and then `wp-env destroy`. Go to your Docker and remove related containers and images and start from fresh.

> Instead of regenerating the product attributes lookup table at this point, how about enabling lookup direct updates? That's what we did in our [pre-existing e2e tests](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/fixtures/fixture-loaders.js#L484-L490) to circumvent the same problem of not having this table populated.

Yes I'v already tried that but it doesn't work. I believe it is because when you import with CSV via the WP importer, that setting is not triggered. I am not sure how the non PW testings are importing the products to know why it works over there.

roykho committed:
Merge branch 'trunk' into add/e2e-tests-attributes-count

roykho commented:
I think I know why you're seeing the 500s and/or 400s. It is because the `Filter by Attributes Block` page wasn't created. The reason why it isn't created successfully is because during import, the product attributes were not created. You can check by going to `http://localhost:8889/wp-admin/edit.php?post_type=product&page=product_attributes` after you have run `npm run env:start`. If it's empty, that is the reason.

Though I am not sure yet why the attributes are not being generated all of a sudden. When I submitted the PR, it was working fine. I am not sure if something changed with the environment like WP version or something else. If I manually create the attributes and page. The tests work fine as shown here:

![Rujyn1.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/13ab0860-c326-422e-ad23-aff19af5cd3a)

Note that even with `trunk`, the attributes are not being generated either...I have to investigate. Something must of change between the time I submitted the PR and now.

roykho committed:
Prevent tests from passing if test page is not loaded

roykho commented:
After some more digging, I found that if you import via the admin UI, it works as expected to create all the product attributes. But when it is imported via the CLI (as configured in the env), it fails to create the product attributes. I don't yet know why though 🤷

roykho committed:
Use WP wrapper to call WC CLI

roykho commented:
@nefeline - ok I think I fixed it. It was because the CLI command didn't use the `wp` cli wrapper. If you don't mind taking a look at this again.

nefeline reviewed (approved):
> Do you have any suggestions so that it is descriptive enough otherwise if we have a fail somewhere, all the descriptions would be the same and you would have to count to know which failed. I had previously thought about 'Filter by Attributes Block - with All products Block 1...2...3'. But it decided against it.

Gotcha! Well, I don't have any strong preferences/suggestions here. If you think there's no way to re-write these phrases without losing their meaning, we can keep them as-is. An example I thought of was:

`should count the products correctly based on the filter (color=blue|query_type_color=or)` 

as a replacement for:

`should show the correct count number of products based on the filter (color=blue|query_type_color=or)`

But that's just a minor/optional suggestion to improve readability while the tests are running.

> I think I know why you're seeing the 500s and/or 400s. It is because the Filter by Attributes Block page wasn't created. The reason why it isn't created successfully is because during import, the product attributes were not created. You can check by going to http://localhost:8889/wp-admin/edit.php?post_type=product&page=product_attributes after you have run npm run env:start. If it's empty, that is the reason.
> Though I am not sure yet why the attributes are not being generated all of a sudden. When I submitted the PR, it was working fine. I am not sure if something changed with the environment like WP version or something else. If I manually create the attributes and page. The tests work fine as shown here
> ok I think I fixed it. It was because the CLI command didn't use the wp cli wrapper. If you don't mind taking a look at this again.

Awesome 🕵️ work! I can confirm all tests are now passing without any server errors:

<img width="1979" alt="Screenshot 2023-05-29 at 18 19 16" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/66e5b020-32b0-4a3c-bf21-b775815450a0">

> Yes I'v already tried that but it doesn't work. I believe it is because when you import with CSV via the WP importer, that setting is not triggered. I am not sure how the non PW testings are importing the products to know why it works over there.

In [this particular](https://github.com/woocommerce/woocommerce-blocks/pull/9500#discussion_r1205980371) case, I believe generating the attribute as soon as the products are created is the ideal path and also faster than having them regenerated later on (after the products were already imported), but unfortunately, I don't have much experience with PW, so I can't think of any immediate solution at the top of my head, but tagging @gigitux and @nielslange (considering your work on pdToLP-wu-p2) in case you have any ideas.

I'm pre-approving the PR as the tests are now passing, but I believe ideally we should optimize the way the meta-lookup table is populated at some point so the tests can run a bit faster.

nielslange reviewed (approved):
> In https://github.com/woocommerce/woocommerce-blocks/pull/9500#discussion_r1205980371 case, I believe generating the attribute as soon as the products are created is the ideal path and also faster than having them regenerated later on (after the products were already imported), but unfortunately, I don't have much experience with PW, so I can't think of any immediate solution at the top of my head, but tagging @gigitux and @nielslange (considering your work on pdToLP-wu-p2) in case you have any ideas.

@nefeline and @roykho Isn't this part work as expected: https://github.com/woocommerce/woocommerce-blocks/blob/45109f4ebcf772a2a3edb1cdf3da01996dcfbd6b/tests/e2e-pw/bin/test-env-setup.sh#L42

---

As for the testing names, below I listed a few suggestions. To increase the readability, I listed them in pairs. Each first snippet shows the current name, each second snippet shows my suggestion. 

```ts
Filter by Attributes Block - with All products Block
```
```
Filter by Attributes Block
```

<br>

```ts
should show the correct count number of products based on the filter (color=blue|query_type_color=or)
```
```
shows correct product count for: color=blue, query_type_color=or
```

<br>

```ts
should show the correct count number of products based on the filter (color=blue,gray|query_type_color=or)
```
```
shows correct product count for: color=blue,gray, query_type_color=or
```

<br>

```ts
should show the correct count number of products based on the filter (color=blue|query_type_color=or|min_price=15|max_price=40)
```
```
shows correct product count for: color=blue, query_type_color=or, min_price=15, max_price=40
```

---

That said, I also reviewed the PR and the E2E tests work as expected. Therefore, I'm also approving this PR. Thanks for working on this, @roykho. 🙌

gigitux commented on tests/e2e-pw/bin/test-env-setup.sh:
```diff
@@ -128,6 +127,16 @@ if [ "$pa_color" ] && [ "$pa_size" ]; then
 	"
 fi
 
+if [ "$pa_color" ]; then
```
@nielslange I think that we should create a script that generates this command for each .txt that is inside the folder bin. What do you think?

nielslange replied:
I'm uncertain if I follow you, @gigitux. The *.txt files inside the bin folder, are the individual blocks, e.g. Active Filters block, All Products blocks, etc. Shouldn't the test products be decoupled from the individual blocks?

gigitux replied:
Sorry if I wasn't clear. As you can see, now we have a specific command for each `.txt` file that is in the `bin` folder. https://github.com/woocommerce/woocommerce-blocks/blob/90760e4be19c5e719d8ae20fcb372d01ce63c191/tests/e2e-pw/bin/test-env-setup.sh/#L113-L153

It is easy to maintain because we have few `.txt` files, but it will not be when there are more. So, I'm suggesting creating a dedicated script that automatically inserts all the posts/pages reading all the `.txt`.

gigitux replied:
Just to clarify, it isn't something that we should do in this PR, but after reading the code it came to my mind!

nielslange replied:
Thanks for clarifying this! I thought you were referring to `if [ "$pa_color" ]; then`, as I see that part highlighted. Automating the loading of the *.txt files definitely makes sense. The current approach looks very redundant.

gigitux replied:
@nielslange, thanks for the feedback! Issue created! https://github.com/woocommerce/woocommerce-blocks/issues/9646

gigitux commented on tests/e2e-pw/tests/attributes-filter/filter-products-by-attributes-count.block_theme.spec.ts:
```diff
@@ -0,0 +1,139 @@
+test.describe( 'Filter by Attributes Block - with All products Block', () => {
+	test.beforeAll( async () => {
+		await cli(
+			`npm run wp-env run tests-cli wp plugin activate woocommerce-blocks`
```
This is something that I noticed and fixed in another PR #9575. We should enable the plugin on setup level (https://github.com/woocommerce/woocommerce-blocks/pull/9575/files#diff-eb99670fee72ab86b8e74c0a34e285c93b76dd5322e9629488ceefd262b33f29R14)

nielslange replied:
@gigitux Somehow, I missed that in #9148, the WooCommerce Blocks plugin had been removed from `.wp-env.json`, see https://github.com/woocommerce/woocommerce-blocks/pull/9148/files#diff-9fd21b09a0144355cfd001e3952767df08b2e04241994470aa1d627f797b171e

Was there a particular reason not to load WooCommerce Blocks by default?

gigitux replied:
@nielslange If I remember well, with the newer version of `@wordpress/env` is necessary to use [`mappings`](https://github.com/woocommerce/woocommerce-blocks/pull/9148/files#diff-9fd21b09a0144355cfd001e3952767df08b2e04241994470aa1d627f797b171eR11-R15) to pass local files to WordPress.

nielslange replied:
I understand the [mappings](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#avoid-activating-plugins-or-themes-on-the-instance) more like something that allows us to install a plugin, but not activate it by default. 

I just changed the `plugins` and `env` sections in `.wp-env.json` to this:
```json
"plugins": [
	"https://downloads.wordpress.org/plugin/woocommerce.latest-stable.zip",
	"https://github.com/WP-API/Basic-Auth/archive/master.zip",
	"https://downloads.wordpress.org/plugin/wordpress-importer.0.8.zip",
	"./tests/mocks/woo-test-helper",
	"."
],
"env": {
	"tests": {
		"mappings": {
			"wp-content/mu-plugins": "./node_modules/@wordpress/e2e-tests/mu-plugins",
			"wp-content/plugins/gutenberg-test-plugins": "./node_modules/@wordpress/e2e-tests/plugins",
			"wp-content/plugins/woocommerce-blocks": ".",
			"wp-cli.yml": "./wp-cli.yml"
		}
	}
},
```

As expected, the E2E tests of this PR passed:

<img width="1512" alt="Screenshot 2023-05-30 at 16 45 54" src="https://github.com/woocommerce/woocommerce-blocks/assets/3323310/cbd20b95-98cb-4966-819e-03da9e8a49c6">

Instead of mapping and later activating WooCommerce Blocks, I'd say it's better to activate the plugin immediately by adding it to the `plugins` section in `.wp-env.json`.

gigitux replied:
>  I'd say it's better to activate the plugin immediately by adding it to the plugins section in .wp-env.json.

Agree! I will create a PR to fix this! Nice catch! Thanks! 💪

nielslange replied:
Thanks for creating a separate PR for that. Feel free to assign me to review it. ;)

gigitux commented on tests/e2e-pw/tests/attributes-filter/filter-products-by-attributes-count.block_theme.spec.ts:
```diff
@@ -0,0 +1,139 @@
+
+		const expectedValues = [ '4', '0', '2', '2', '0' ];
+
+		const lists = await page.$$(
```
We should not use `$$` on playwright. We should use locators that are the central piece of Playwright's auto-waiting and retry-ability. ([more details](https://playwright.dev/docs/api/class-locator)) - [internal documentation](https://github.com/woocommerce/woocommerce-blocks/blob/9c33c0819e2de8c1390ee560fdaf35df804d4035/docs/contributors/contributing/e2e-guidelines.md/#L42-L44).

gigitux reviewed (changes requested):
Thanks for working on E2E tests. I left some feedback!

gigitux commented on .wp-env.json:
```diff
@@ -26,6 +26,7 @@
 	"config": {
 		"JETPACK_AUTOLOAD_DEV": true,
 		"SCRIPT_DEBUG": false,
+		"WP_DEBUG": true,
```
Is there a reason behind this?

roykho replied:
Yes, sometimes when testing back to back, changes are picked up from cache and gives false negatives. But if you think this causes other issues, I can remove it.

roykho replied:
I've decided to remove it for now because in a "normal" automated testing situations where we're not actually debugging it by testing back to back, this setting isn't needed.

nefeline commented:
Regarding:

> @nefeline and @roykho Isn't this part work as expected:

There's a difference between populating the lookup table at the same time new products are created versus regenerating the lookup tables later on via a separate task. This is an extra step we didn't have to run while working on Puppeteer, but now we do it on Playright. 

If folks don't have any ideas to solve this at the top of their head, my recommendation is to create a separate issue so we can investigate later on, as performance will certainly be impacted.

roykho commented:
> @nefeline and @roykho Isn't this part work as expected:

So I believe this does not include `wp_wc_product_attributes_lookup` table. That's why they are separate buttons in the status->tools section.

roykho committed:
Refactor to use more of PW methods

roykho committed:
Use existing active filters block post for testing

roykho commented:
@nefeline @nielslange @gigitux 

Thanks for the feedbacks! I've updated the PR to address as much as possible. Here are the notable changes from the original PR.

* I refactor the use of generic `page.$ and page.$$` to use PW locators instead.
* I removed the added post creation of `Filter by Attributes Block` because I realize we could just use the existing `Active Filters Block` post.
* I removed the `WP_DEBUG` statement as normally this is not needed during automated testing.
* I shorten the test titles a bit. They are still long but much shorter.
* I utilize the method @nielslange mentioned to activate the Blocks plugin in `.wp-env.json`.

With that, if ya'll don't mind giving it another look. Thanks!

nefeline commented:
Thanks for addressing the feedback and for shortening the tests description as suggested. 🙏 

My only remaining concern is still related to https://github.com/woocommerce/woocommerce-blocks/pull/9500#discussion_r1205980371 : since none of us involved in this discussion had any ideas/suggestions to update this configuration at the top of our heads, my recommendation is to create an issue so we can investigate later on (during a cooldown).

All tests are passing, but I'm deferring the approval to Luigi & Niels since they requested additional changes regarding the tests setup 👍 .

roykho commented:
@nefeline - created a separate issue https://github.com/woocommerce/woocommerce/issues/42355

gigitux commented on tests/e2e-pw/tests/attributes-filter/filter-products-by-attributes-count.block_theme.spec.ts:
```diff
@@ -0,0 +1,112 @@
+};
+
+test.describe( 'Filter by Attributes Block - with All products Block', () => {
+	test.beforeEach( async ( { page } ) => {
+		await prepareAttributes( page );
+	} );
```
Two observations:
- Is it correct that this function runs before each test? Isn't it better `beforeAll`?
- At the same, this is a global side effect that could influence other E2E tests. What do you think about moving this function to the `global-setup.ts` file? (we can refactor this in #9656)

roykho replied:
Ok I have moved the `prepareAttributes` function to the global setup.

gigitux reviewed (commented):
Thanks for addressing the feedback! I'm not sure that `prepareAttributes` function runs in the right place. What do you think?

gigitux commented:
> @nefeline - created a separate issue woocommerce/woocommerce#42355

Thanks! I added this issue into the scope of woocommerce/woocommerce#42413 🚀

roykho committed:
Move prepareAttributes function to global setup

nielslange commented on tests/e2e-pw/global-setup.ts:
```diff
@@ -75,6 +75,52 @@ const loginAsCustomer = async ( config: FullConfig ) => {
+	 * duplicated as we need to run the cron task twice as
+	 * we need to process more than 1 batch of items.
+	 */
+	await cli(
+		`npm run wp-env run tests-cli wp action-scheduler run --hooks="woocommerce_run_product_attribute_lookup_regeneration_callback"`
+	);
+
+	await cli(
+		`npm run wp-env run tests-cli wp action-scheduler run --hooks="woocommerce_run_product_attribute_lookup_regeneration_callback"`
+	);
```
While doing the review, this section caught my attention every time I read over it. It's OK to keep it for now, but I wonder if there isn't an alternative way to accomplish this.

roykho replied:
Yeah there is already a separate PR to track this. I personally tried many different ways to get around this but to no avail. I think the only way is to make some changes upstream.

nielslange reviewed (approved):
Thanks for addressing all changes, @roykho. I left a comment on a potential future improvement, but that's out of the scope for this PR.
